### PR TITLE
[AutoDiff] Fix `@differentiable` attribute where clause printing.

### PR DESCRIPTION
**Pull request by dan-zheng**
Only print where clause requirements that are not satisfied by the original
function's generic signature. This improves diagnostics and reduces noise.

Resolves [TF-371](https://bugs.swift.org/browse/TF-371).

### CODE REVIEW

**Comment by dan-zheng** [comment]
Note: merge #23383 first because it contains a `FIXME(TF-371)`, which is resolved by this PR.

File: lib/AST/Attr.cpp [hunk 1]
```diff
@@ -406,7 +408,19 @@ static void printDifferentiableAttrArguments(
         return genericEnv->getSugaredType(Ty);
       };
     }
-    interleave(attr->getRequirements(), [&](Requirement req) {
+    // Filter out requirements satisfied by original function's generic
+    // signature. They should not be printed.
+    auto requirementsToPrint =
+      makeFilterRange(attr->getRequirements(), [&](Requirement req) {
```
Comment (rxwei) [inline lib/AST/Attr.cpp]: Indent by 4 here.

**Review by rxwei — APPROVED** [review]

File: lib/AST/Attr.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/AST/Attr.cpp]: Thanks! Done.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
Merging to unblock progress. All macOS tests pass locally.

**Pull request merged**